src/content/blog/first-four-val-town-runtimes.mdx

```diff
@@ -0,0 +1,247 @@
+---
+title: The first four Val Town runtimes
+generated: 1701894028894
+description: From vm to vm2 to workers to processes, the long journey to how we run vals today
+pubDate: February 5, 2024
+heroImage: ./first-four-val-town-runtimes/timeline.png
+author: Tom MacWright
+---
+
+import { Image } from "astro:assets";
+import memoryImage from "./first-four-val-town-runtimes/memory.png";
+
+One of the most important things that Val Town does is **run untrusted code**.
+Our whole premise requires us to take the TypeScript code that our community
+writes and run it on a server, safely, quickly, and flexibly.
+
+#### Safety
+
+The default way to run arbitrary code is "dangerously". Modules in Node.js have unfettered
+access to your system's environment variables, filesystem, network, and much
+more. This access is [routinely abused in supply-chain attacks](https://socket.dev/blog/inside-node-modules).
+The easiest way to run an arbitrary string of code is with the `eval()` method,
+which doesn't provide any kind of sandbox protections, so it can easily become
+an attack vector.
+
+Security generally means isolation and containment:
+
+- Vals need to be isolated from the system they're running on: they shouldn't
+  be able to read from the filesystem or learn about environment variables.
+- Vals should be isolated from each other: you shouldn't be able to affect
+  anyone else's val's behavior.
+
+#### Speed
+
+On the other hand, if you were to make every Val run in a full-fledged Docker
+container, it'd be tremendously slow. You're booting up a whole Linux machine
+with all the bells and whistles just to evaluate a few lines of code. It'd be
+[pretty secure](https://docs.docker.com/engine/security/#conclusions), but
+pretty slow.
+You could create a pool of Docker containers, but the memory and devops overhead
+would be significant.
+
+#### Flexibility
```

Comment:
I might entitled this section "NPM" or "libraries" or "compatibility" or something like that